### PR TITLE
[10120] Update the PR template with the merge commit message.

### PR DESCRIPTION
**Pull request by adiroiban**
## Scope and purpose

The commit message is extracted from here

https://twistedmatrix.com/trac/wiki/ReviewProcess#Authors:Howtomergethechangetotrunk

------

It updates the contribution list as I think that tox -e lint is no longer needed... I hope that by now peole will have pre-comm it installed .. if not, we should update the docs to encorage usage of pre-commit.

If tox -e lint is not called, a check will fail anyway.

----

It add info about adding the trac ticket in the PR title.
In this way, I hope that in the future we can add a robot that could automatically generate the merge message.

Author can be extracted from the list of committers.
Reviewers from the list of reviewers that have approved the PR

For now, only add ticket ID and see how it goes.

In the future, I hope that we can have the full newsfragment desciption in the PR title so that we can also auto-generate the
newsfragment file from the info found in the PR.

------------

It also ask for adding the Twisted team to the list of reviewers and in this way it will trigger the GitHub notifications. 
I hope that in this way the PR will be more visible.

And in the future, when the Twisted team is assigned, I hope that we can have a bot that will automatically assign a "review needed" tag to this PR for easy filtering.

## Contributor Checklist:

* [x] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/10120
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [NA] I have updated the automated tests.
* [x] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.

Below is the template for easy rendered reviewing

----------------------

----------------------------

<!-WILL-BE-COMMENT
## Remove this section

Have a look at [our developer documentation](https://twistedmatrix.com/trac/wiki/TwistedDevelopment#SubmittingaPatch) before submitting your Pull Request.

Note that the Trac ticket, news fragment, and review submission portions of this process apply to *all* pull requests, no matter how small;
if you don't do them, it's likely that nobody will even notice your PR needs a review.
WILL-BE-COMMENT->

## Scope and purpose

Add a few words about why this PR is needed and what is its scope.

Add any comments about trade-offs (if any) made in this PR and the reasoning behind them.

Add mentions of things that are not covered here and are planed to be done
in separate PRs.

## Contributor Checklist:

* [ ] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/
* [ ] I have created a newsfragment in src/twisted/newsfragments/ (see: [News files](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles))
* [ ] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [ ] I have updated the automated tests and checked that all checks for the PR are green.
* [ ] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [ ] I have added `twisted/twisted-contributors` teams to the PR `Reviewers`.
* [ ] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.

```
Merge pull request #123 from twisted/4356-branch-name-with-trac-id

Author: <github_username>, <github_usernames_if_more_authors>
Reviewer: <github_username>, <github_usernames_if_more_reviewers>
Fixes: ticket:<trac_ticket_number>, ticket:<another_if_more_in_one_PR>

Long description providing a summary of these changes.
(as long as you wish)
```

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE [hunk 1]
```diff
@@ -2,21 +2,42 @@
...
+(as long as you wish)
+```
+
+
+## Remove this section
+
+Have a look at [our developer documentation](https://twistedmatrix.com/trac/wiki/TwistedDevelopment#SubmittingaPatch) before submitting your Pull Request.
+
+Note that the Trac ticket, news fragment, and review submission portions of this process apply to *all* pull requests, no matter how small;
+if you don't do them, it's likely that nobody will even notice your PR needs a review.
```
Comment (altendky) [inline .github/PULL_REQUEST_TEMPLATE]: Some projects put this sort of stuff in a comment so it doesn't have to be deleted to not spam everyone looking at the issue.  Regardless, it seems like this warning should go at the very beginning.  It is exactly for people that don't follow the checklist, not for people that do carefully follow the checklist and read every detail down to line 43.

File: .github/PULL_REQUEST_TEMPLATE [hunk 2]
```diff
@@ -2,21 +2,42 @@
 
 Add a few words about why this PR is needed and what is its scope.
 
+Add any comments about trade-offs (if any) made in this PR and the reasoning behind them.
+
 Add mentions of things that are not covered here and are planed to be done
 in separate PRs.
```
Comment (altendky) [inline .github/PULL_REQUEST_TEMPLATE]:
```suggestion
Add mentions of things that are not covered here and are planned to be done in separate PRs.
```

File: .github/PULL_REQUEST_TEMPLATE [hunk 3]
```diff
@@ -2,21 +2,42 @@
...
+* [ ] The title of the PR starts with the associated Trac ticket number (without the `#` character).
+
+* [ ] I have updated the automated tests and checked that all checks for the PR are green.
+
 * [ ] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
+* [ ] I have added `twisted/twisted-contributors` teams to the PR `Reviewers`.
+
+* [ ] The merge commit will use the below format
+  The first line is automatically generated by GitHub based on PR ID and branch name.
+  The other lines generated by GitHub should be replaced.
```
Comment (altendky) [inline .github/PULL_REQUEST_TEMPLATE]: The blank lines seem to affect the formatting.

```suggestion
* [ ] The title of the PR starts with the associated Trac ticket number (without the `#` character).
* [ ] I have updated the automated tests and checked that all checks for the PR are green.
* [ ] I have submitted the associated Trac ticket for review by adding the word `review` to the keywords field in Trac, and putting a link to this PR in the comment; it shows up in https://twisted.reviews/ now.
* [ ] I have added `twisted/twisted-contributors` teams to the PR `Reviewers`.
* [ ] The merge commit will use the below format
  The first line is automatically generated by GitHub based on PR ID and branch name.
  The other lines generated by GitHub should be replaced.
```

**Review by altendky — CHANGES_REQUESTED** [review]
How about requesting that people create PRs as drafts?  Then they can complete the checklist, including verifying themselves that CI checks pass, prior to removing the draft mark.

---

> In the future, I hope that we can have the full newsfragment desciption in the PR title so that we can also auto-generate the newsfragment file from the info found in the PR.

I'm not convinced that the PR title is a good place to write possibly multiple lines worth of reStructuredText.  Also, to get that into a file that'll be able to be rendered by RTD while doing the doc builds doesn't seem great as it will be either injecting automatic commits on a regular basis into the user's PR or otherwise getting way more complicated RTD integration.

---

> It also ask for adding the Twisted team to the list of reviewers and in this way it will trigger the GitHub notifications.  I hope that in this way the PR will be more visible.
>
> And in the future, when the Twisted team is assigned, I hope that we can have a bot that will automatically assign a "review needed" tag to this PR for easy filtering.

I certainly haven't thought through a full solution but I would tend to hope we could minimize the GitHub automation?  An alternative to keeping a tag in sync might be to guide people towards a search including `draft:false review:required`.  For the universal automatic review assignment, I see value but I have some vague concern that it just ends up as spam in some cases too.

**Comment by adiroiban** [comment]
Thanks @altendky  for the review.

I will address your comments and send it back to review :)

> How about requesting that people create PRs as drafts? Then they can complete the checklist, including verifying themselves that CI checks pass, prior to removing the draft mark.

For me this is just extra mouse/trackpad work.

After removing from draft you will still need to add the Twisted team to reviewers list so that the team is notified that the PR is ready for review

so I am -1 for draft

-----------

> I'm not convinced that the PR title is a good place to write possibly multiple lines worth of reStructuredText.

Agree.

----------

> I certainly haven't thought through a full solution but I would tend to hope we could minimize the GitHub automation? 

Agree... not finding a process without automation might not be that easy :)

> An alternative to keeping a tag in sync might be to guide people towards a search including `draft:false review:required`.

The problem with tags, is that only repo members can modify the tags... so one time contributors can only touch the Reviewers field.

>  For the universal automatic review assignment, I see value but I have some vague concern that it just ends up as spam in some cases too.

We can create a separate twisted-reviewers team

File: .github/PULL_REQUEST_TEMPLATE [hunk 4]
```diff
@@ -1,22 +1,41 @@
...
 
-## Remove this section
+Add a few words about why this PR is needed and what is its scope.
 
-Please have a look at [our developer documentation](https://twistedmatrix.com/trac/wiki/TwistedDevelopment#SubmittingaPatch) before submitting your Pull Request.
+Add any comments about trade-offs (if any) made in this PR and the reasoning behind them.
 
-Please note that the trac ticket, news fragment, and review submission portions of this process apply to *all* pull requests, no matter how small; if you don't do them, it's likely that nobody will even notice your PR needs a review.
+Add mentions of things that are not covered here and are planed to be done
+in separate PRs.
```
Comment (altendky) [inline .github/PULL_REQUEST_TEMPLATE]: Fix semantic newlines while we're in here?

```suggestion
Add mentions of things that are not covered here and are planed to be done in separate PRs.
```

**Review by altendky — APPROVED** [review]

File: .github/PULL_REQUEST_TEMPLATE [hunk 5]
```diff
@@ -1,22 +1,41 @@
...
 
 
 ## Contributor Checklist:
 
 * [ ] The associated ticket in Trac is here: https://twistedmatrix.com/trac/ticket/<!-- Create a new one at https://twistedmatrix.com/trac/newticket and replace this comment with the ticket number. -->
-* [ ] I ran `tox -e lint` to format my patch to meet the [Twisted Coding Standard](https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html)
```
Comment (wsanchez) [inline .github/PULL_REQUEST_TEMPLATE:19]: Most of us use `pre-commit`, but one-off contributors, who benefit most from this checklist, perhaps do not. You're right that CI catches it, but I don't see the harm in keeping this here.

**Review by wsanchez — APPROVED** [review]

File: .github/PULL_REQUEST_TEMPLATE (hunk 5 above)
Comment (twm) [reply .github/PULL_REQUEST_TEMPLATE:19]: FWIW, I was pretty confused when I pulled down trunk and `tox -e black-reformat` no longer worked. This sort of workflow change should be announced on the mailing list. As the contributor docs weren't updated either I only learned what I was supposed to do from this checklist item.

Comment (adiroiban) [reply .github/PULL_REQUEST_TEMPLATE:19]: Thanks Tom for the feedback. Agree with the confusion. 

I see twisted dev workflow as a wild wild west. Everyone has it's style.
For example, I don't like tox for local dev and TDD  and don't use `tox -e lint`

So for now, the only gold standard in terms of workflow are the required CI checks.

**Pull request merged**